pandas/core/groupby.py

```diff
@@ -3740,9 +3740,21 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        indexer = self.grouper.result_index.get_indexer(index)
+        result = result.reindex(index=indexer, copy=False)
+        result.index = range(len(result))
+        # Update values of grouper columns with index levels
+        for i in range(index.nlevels):
```

Comment:
Perhaps, a bit safer is something like this:
```python
for i in range(len(groupings)):
    if groupings[i].in_axis:
        result[groupings[i].name] = index.get_level_values(i)
```
But I don't have a clue what else I could do with it.